photon-lib/src/main/scala/com/linkedin/photon/ml/hyperparameter/estimators/GaussianProcessEstimator.scala

```diff
@@ -0,0 +1,148 @@
+ * @param monteCarloNumBurnInSamples the number of samples to draw during the burn-in phase of kernel parameter
+ *   estimation
+ * @param monteCarloNumSamples the number of samples to draw for estimating kernel parameters
+ */
+class GaussianProcessEstimator(
+    kernel: Kernel = new RBF,
+    normalizeLabels: Boolean = false,
+    predictionTransformation: Option[PredictionTransformation] = None,
+    monteCarloNumBurnInSamples: Int = 100,
+    monteCarloNumSamples: Int = 100,
+    seed: Long = System.currentTimeMillis) {
+
+  /**
+   * Produces a Gaussan Process regression model from the input features and labels
+   *
+   * @param x the observed features
+   * @param y the observed labels
+   * @return the estimated model
+   */
+  def fit(x: DenseMatrix[Double], y: DenseVector[Double]): GaussianProcessModel = {
+    require(x.rows > 0 && x.cols > 0, "Empty input.")
+    require(x.rows == y.length, "Training feature sets and label sets must have the same number of elements")
+
+    // Normalize labels
+    val (yTrain, yMean) = if (normalizeLabels) {
+      val m = mean(y)
+      (y - m, m)
+    } else {
+      (y, 0.0)
+    }
+
+    val kernels = estimateKernelParams(x, yTrain)
+    new GaussianProcessModel(x, yTrain, yMean, kernels, predictionTransformation)
+  }
+
+  /**
+   * Estimates kernel parameters by sampling from the kernel parameter likelihood function
+   *
+   * We assume a uniform prior over the kernel parameters $\theta$ and observed features $x$, therefore:
+   *
+   *   $l(\theta|x,y) = p(y|theta,x) \propto p(theta|x,y)$
+   *
+   * Since the slice sampling algorithm requires that the function be merely proportional to the target distribution,
+   * sampling from this function is equivalent to sampling from p(\theta|x,y). These samples can then be used to compute
+   * a Monte Carlo estimate of the response for a new query point $q'$ by integrating over values of $\theta$:
+   *
+   *   $\int r(x', \theta) p(\theta) d\theta$
+   *
+   * In this way we (approximately) marginalize over all $\theta$ and arrive at a more robust estimate than would be
+   * produced by computing a maximum likelihood point estimate of the parameters.
+   *
+   * @param x the observed features
+   * @param y the observed labels
+   * @return a collection of covariance kernels corresponding to the sampled kernel parameters
+   */
+  protected[estimators] def estimateKernelParams(
+      x: DenseMatrix[Double],
+      y: DenseVector[Double]): List[Kernel] = {
+
+    val logp = (theta: DenseVector[Double]) => logLikelihood(x, y, theta)
+    val sampler = new SliceSampler(logp, range = kernel.getParamBounds, seed = seed)
+
+    // Sampler burn-in. Since Markov chain samplers like slice sampler exhibit serial dependence between samples, the
+    // first n samples are biased by the initial choice of parameter vector. Here we perform a "burn in" procedure to
+    // mitigate this.
```

Comment:
There is some discussion here: https://en.wikipedia.org/wiki/Metropolis%E2%80%93Hastings_algorithm

The basic problem is that the initial choice introduces some sampling bias. Even if you choose the initial sample uniformly at random, it could land in very unlikely areas of the function and produce skewed samples in the first few iterations.